### PR TITLE
Properfly fixing the seed to reproduce experiments

### PR DESCRIPTION
**Pull request by adrianjav**
I've made some changes to enable true reproducibility based on another project where I had to enable that feature. The changes:
* Fixed the seed for numpy as well (which is used in transformer, p.e.)
* Set the CUDNN backend to deterministic mode, otherwise you may find small variations on each experiment.
* Now the random seed is always set if provided.
* Some typos changed.

Hope it helps.

### CODE REVIEW

**Comment by vince62s** [comment]
can you please point where we use numpy random functions ? I was able to have reproducibility so far, so I am curious. thanks

**Comment by adrianjav** [comment]
there isn't any usage of those at the moment, but the code does use numpy already and it could use shadowed random functions in the future, for example, in a function that has to decide on ties. I like to fix the seed if I make use of the library.

Making cudnn run on deterministic mode is essential though. I had experiments varying in the actual numeric outputs (not in the general behaviour) and it was due to cudnn.

**Comment by bpopeters** [comment]
@vince62s @srush I hardly think that this is overweighting the code. The changes are small, the logic is clear, and reproducibility is obviously necessary. I've been having trouble with reproducibility as well.

(that said, I can't speak to the particulars of this PR because it's not my area of expertise)

**Comment by vince62s** [comment]
It's fine, just saying that we do not use numpy random things in the code right now, so we already have reproducibility  (maybe except the fix for cudnn deterministic set up). I have tested it when refactoring a few weeks ago.

**Comment by bpopeters** [comment]
Do we? I tested this morning and I can't replicate results across multiple runs with the same seed even just training on the CPU.

**Comment by adrianjav** [comment]
@bpopeters have you tested them on GPU as well? At the moment the seed is only fixed if you're using the GPU

**Comment by bpopeters** [comment]
It isn't reproducible with the GPU either. Is there a good reason why you can't set the seed using the CPU?

**Comment by adrianjav** [comment]
Not that I'm aware of. That's changed in this PR as well, so the seed is fixed in the CPU case.
Note that the changes are just for the single-gpu case

**Comment by vince62s** [comment]
@bpopeters were you on master when you tried ? can you share your command line please ?
because I had hard time to fix the reproducibility, so I hope nothing broke it.

**Comment by vince62s** [comment]
guys,
with this command line
$my_python preprocess.py -train_src data/src-train.txt -train_tgt data/tgt-train.txt -valid_src data/src-val.txt -valid_tgt data/tgt-val.txt -save_data data/data -src_vocab_size 1000 -tgt_vocab_size 1000 -share_vocab


$my_python train.py -data data/data -save_model /tmp/tmp -batch_type tokens -batch_size 1024 -accum_count 4 \
 -layers 4 -rnn_size 512 -word_vec_size 512 -encoder_type transformer -decoder_type transformer -share_embedding \
 -train_steps 10000 -gpuid 0 -max_generator_batches 4 -dropout 0.1 -normalization tokens -seed 7070 \
 -max_grad_norm 0 -optim adam -decay_method noam -learning_rate 2 -label_smoothing 0.1 \
 -position_encoding -param_init 0 -warmup_steps 1000 -param_init_glorot -adam_beta2 0.998


I can replicate the exact same log for the first 10 iterations while if I change the seed I get a different log for each seed.

**Comment by adrianjav** [comment]
I've spent some time today trying to reproduce any case where two clone models differ but I have achieved yet. On previous projects I had that problem, but it seems that is related with just some functions of cuDNN, specially the ones related with CNNs. Also, whether cuDNN decides to use those kinds of non-deterministic functions or not depends on the problem size.

Either way, there are plenty of issues addressing this very same problem:
* deeplearning4j/deeplearning4j#5089 
* pytorch/pytorch#114
* tensorflow/tensorflow#12871

**Comment by vince62s** [comment]
Just to be clear, I'm fine with this line:
torch.backends.cudnn.deterministic = True
So I suggest we commit now and we postpone the numpy one when really needed.
ok ?

**Comment by adrianjav** [comment]
I agree. Should I commit on this branch removing the numpy thing? Or can you cherry pick the lines to be pulled?
Also, this PR also changes the code to fix the seed when using the CPU. That's fine, right?

**Comment by vince62s** [comment]
I thought it was only needed for GPU. But that's fine I guess.

**Comment by adrianjav** [comment]
Nah, PRNG's seed needs to be fixed no matter the device you are using.

I've removed the numpy dependencies, so it's fine to go.
